src/cheevos/Cheevos.cpp

```diff
@@ -0,0 +1,285 @@
+/*
+ *  This file is based on https://github.com/libretro/RetroArch/blob/96c5f5dfb07454c972c50838815330382d6b1911/cheevos/fixup.c
+ *
+ *  Copyright (C) 2014-2020 Team Kodi (https://kodi.tv)
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSE.md for more information.
+ */
+
+#include "Cheevos.h"
+
+#include "libretro/LibretroEnvironment.h"
+#include "libretro/MemoryMap.h"
+#include "libretro/libretro.h"
+#include "rcheevos/rconsoles.h"
+#include "rcheevos/rhash.h"
+#include "rcheevos/rurl.h"
+
+using namespace LIBRETRO;
+
+namespace
+{
+constexpr int HASH_SIZE = 33;
+constexpr int URL_SIZE = 512;
+constexpr int RICH_PRESENCE_EVAL_SIZE = 512;
+constexpr int POST_DATA_SIZE = 1024;
+} // namespace
+
+CCheevos::CCheevos()
+{
+  rc_runtime_init(&m_runtime);
+}
+
+void CCheevos::Initialize()
+{
+  rc_runtime_init(&m_runtime);
+}
+
+void CCheevos::Deinitialize()
+{
+  rc_runtime_destroy(&m_runtime);
+}
+
+void CCheevos::ResetRuntime()
+{
+  rc_runtime_reset(&m_runtime);
+  m_richPresence =
+      rc_parse_richpresence(m_richPresenceBuffer.data(), m_richPresenceScript.c_str(), NULL, 0);
+}
+
+CCheevos& CCheevos::Get(void)
+{
+  static CCheevos _instance;
+  return _instance;
+}
+
+bool CCheevos::GenerateHashFromFile(std::string& hash,
+                                    unsigned int consoleID,
+                                    const std::string& filePath)
+{
+  char _hash[HASH_SIZE] = {};
+
+  int res = rc_hash_generate_from_file(_hash, consoleID, filePath.c_str());
+  hash = _hash;
```

Comment:
`hash` should only by set if `res` == `RC_OK`. However, on the note of checking return value, here's a snippet of (maybe outdated) documentation I found:

---
### Return value

If the function succeeds, the return value is RC_OK. Otherwise, the error code can be converted to a string using [rc_error_str](https://github-wiki-see.page/m/RetroAchievements/rcheevos/wiki/rc_error_str).

*    RC_INVALID_STATE - one or more required parameters was not provided.
*    RC_OUT_OF_MEMORY - enough memory could not be allocated to complete the operation.
---


To be fully correct, we should handle all pre-conditions and post-conditions when calling into an external library. To satisfy post-conditions, this means checking the result and logging errors.

Reply:
The comments above the function state:
```
/* generates a hash from a file.
   * returns non-zero on success, or zero on failure.
   */
```
That's why I didn't use the `RC_OK` you suggested in the previous comment, it was a bit confusing to check if the result is different from the OK value my thoughts when reading that were that if `CCheevos::GenerateHashFromFile` returns true it means it failed